fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/IndexOrphanValidation.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * IndexOrphanValidation.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *  
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *  
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at 
+ *  
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *  
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.metadata;
+
+import com.apple.foundationdb.API;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.IndexEntry;
+import com.apple.foundationdb.record.IndexScanType;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.ScanProperties;
+import com.apple.foundationdb.record.TupleRange;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseRunner;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordContext;
+import com.apple.foundationdb.record.provider.foundationdb.IndexMaintainer;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Validate entries in the index. It scans the index and checks if the record associated with each index entry exists.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class IndexOrphanValidation {
+
+    private IndexOrphanValidation() {}
+
+    public static CompletableFuture<List<IndexEntry>> validate(@Nonnull FDBDatabaseRunner runner, @Nonnull IndexMaintainer indexMaintainer) {
+        return validate(runner, indexMaintainer, ScanProperties.FORWARD_SCAN);
+    }
+    public static CompletableFuture<List<IndexEntry>> validate(@Nonnull FDBDatabaseRunner runner,
```

Comment:
Missing a blank line separating the methods. It also might be nice to have Javadocs on these two methods.

It's also a little weird that they return a list rather than cursors. It looks like this is designed to handle continuations transparently (which is fine), but a cursor still seems like a more natural return type to me (but what do I know?). Then the user can call `.asList` on the cursor if they really want a list.

Maybe it's fine, though. (This is an experimental API, after all.)

Reply:
Blank line added. Will add Javadocs when the APIs are consolidated.

I'm going to talk about the APIs with @scgray later. Another API related issue is providing an interface for `ExecuteProperties` or not.

Reply:
That sounds reasonable to me.

Reply:
Now that I look at this again, I think this could be distilled into a generic cursor, similar to the `ChainedCursor`. But like a `ChainedRecordCursor<T>.  For example:

    RecordCursor<IndexEntry> cursor = new ChainedRecordCursor<>(
        FDBDatabaseRunner runner,
        (context, continuation, scanProperties) -> 
               indexMaintainer.validateOrphanEntries(IndexScanType.BY_VALUE, TupleRange.ALL, continuation, scanProperties),
        limit);

this would actually be pretty useful to have and would basically do away with the need for this class at all.

If you go this route, I would probably create that cursor as a separate PR, then pull it into this one.